docs/advice.org

```diff
@@ -0,0 +1,859 @@
+(import-macros {: defn
+                : defadvice} :lib.advice.macros)
+
+(defn original-fn
+      [x y z]
+      "docstr"
+      true)
+
+(original-fn 1 2 3)
+;; => true
+
+(defadvice advice-fn
+           [x y z]
+           :after-until original-fn
+           "After-until original-fn"
+           (+ x y z))
+
+(original-fn 1 2 3)
+;; => true ;; original-fn returned truthy vaue, advice-fn not called
+#+end_src
+
+** filter-args
+
+The advising function is called with the args provided by the caller,
+it must return a table list of args to apply to the original function.
+It transforms arguments, similar to around but without having access to the original.
+
+Behavior:
+
+#+begin_src fennel
+(fn [...]
+  (original-fn (table.unpack (advice-fn (table.unpack [...])))))
+#+end_src
+
+Example:
+
+#+begin_src fennel
+(import-macros {: defn
+                : defadvice} :lib.advice.macros)
+
+(defn original-fn
+      [x y z]
+      "docstr"
+      (+ x y z))
+
+(original-fn 1 2 3)
+;; => 6
+
+(defadvice advice-fn
+           [x y z]
+           :filter-args original-fn
+           "filter-args original-fn"
+           [(* x 2) (* y 2) (* z 2)])
+
+(original-fn 1 2 3)
+;; => 10 ;; Values returned by advice-fn applied to original-fn
+#+end_src
+
+** filter-return
+
+The advising function is called with the return value of the original
+function. It may transform the return value and return the transformed
+value to the caller. It is also similar to around but without access
+to the original.
+
+Behavior:
+
+#+begin_src fennel
+(fn [...]
+  (advice-fn (original-fn (table.unpack [...]))))
+#+end_src
+
+Example:
+
+#+begin_src fennel
+(import-macros {: defn
+                : defadvice} :lib.advice.macros)
+
+(defn original-fn
+      [x y z]
+      "docstr"
+      (+ x y z))
+
+(original-fn 1 2 3)
+;; => 6
+
+(defadvice advice-fn
+           [sum]
+           :filter-return original-fn
+           "filter-return original-fn"
+           (* sum 2))
+
+(original-fn 1 2 3)
+;; => 12 ;; Return value of original-fn passed to advice-fn
+#+end_src
+
+
+* Advising Targets and Order
+
+** Function references or strings
+
+The add advice APIs accept both a target function or the unique key
+pointing to an advisable function entry. Only functions defined with
+=defn=, =afn=, or =make-advisable= are supported.
+
+For example, if this fennel code was in the []:
+
+#+begin_src fennel
+(import-macros {: defn} :lib.advice.macros)
+
+(defn defn-func-2
+      [x y z]
+      "docstr"
+      "default")
+
+(print defn-func-2.key)
+#+end_src
+
+It would print the following:
+
+#+begin_src fennel
+"test/advice-test/defn-func-2"
+#+end_src
+
+That key is a unique pointer to an advisable function. It can be
+passed as the target to both the =defadvice= macro and =add-advice=
+function. It is always calculated from the =~/.hammerspoon= root, if you
+are creating advisable functions within your =~/.spacehammer= directory,
+the keys will start with ="spacehammer"=.
```

Comment:
Seems github doesn't render this properly: 
![Advising_System_by_eccentric-j_·_Pull_Request__107_·_agzam_spacehammer](https://user-images.githubusercontent.com/497310/132996308-b22bb376-d68b-4d11-b5e6-bcf4d7d5b8da.png)